File: sdk/kotlin/src/main/kotlin/io/qail/QailClient.kt
package io.qail

import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.plugins.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.plugins.websocket.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.serialization.kotlinx.json.*
import io.ktor.websocket.*
import kotlinx.coroutines.*
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonPrimitive
import java.net.URLEncoder
import java.nio.charset.StandardCharsets

// ─── Configuration ──────────────────────────────────────────────────

enum class WebSocketAuthMode {
    NONE,
    HEADER,
    QUERY,
}

/**
 * Configuration for the Qail client.
 *
 * @param url Gateway base URL (e.g. `https://engine.example.com`)
 * @param token Static JWT token
 * @param tokenProvider Async token provider for refreshable auth
 * @param headers Additional default headers
 * @param timeoutMs Request timeout in milliseconds (default: 30000)
 * @param httpClient Custom Ktor HttpClient (optional)
 */
data class QailConfig(
    val url: String,
    val token: String? = null,
    val tokenProvider: (suspend () -> String)? = null,
    val headers: Map<String, String> = emptyMap(),
    val timeoutMs: Long = 30_000,
    val wsAuthMode: WebSocketAuthMode = WebSocketAuthMode.HEADER,
    val wsTokenQueryParam: String = "access_token",
    val httpClient: HttpClient? = null,
)

// ─── Client ─────────────────────────────────────────────────────────

/**
 * Qail Gateway client for Kotlin / Android.
 *
 * Provides fluent query builders, raw DSL execution, and utility
 * endpoints — mirroring the TypeScript and Swift SDK surface area.
 *
 * ```kotlin
 * val qail = QailClient(QailConfig(
 *     url = "https://engine.example.com",
 *     token = "my-jwt"
 * ))
 *
 * // Fluent query
 * val users = qail.from<User>("users")
 *     .where("active", FilterOp.EQ, "true")
 *     .limit(10)
 *     .all()
 *
 * // Raw DSL
 * val res = qail.query<User>("get users fields id, name limit 10")
 * ```
 */
class QailClient(@PublishedApi internal val config: QailConfig) {

    @PublishedApi
    internal val baseUrl: String = config.url.trimEnd('/')

    @PublishedApi
    internal val json: Json = Json {
        ignoreUnknownKeys = true
        isLenient = true
    }

    @PublishedApi
    internal val client: HttpClient = config.httpClient ?: HttpClient {
        install(ContentNegotiation) {
            json(this@QailClient.json)
        }
        install(HttpTimeout) {
            requestTimeoutMillis = config.timeoutMs
        }
    }

    // ── Query builder entry points ──────────────────────────────────

    /** Start a SELECT query on a table. */
    inline fun <reified T> from(table: String): SelectBuilder<T> =
        SelectBuilder(this, table, T::class.java)

    /** Start an INSERT on a table. */
    inline fun <reified T> into(table: String): InsertBuilder<T> =
        InsertBuilder(this, table, T::class.java)

    /** Start an UPDATE on a table. */
    inline fun <reified T> update(table: String): UpdateBuilder<T> =
        UpdateBuilder(this, table, T::class.java)

    /** Start a DELETE on a table. */
    fun delete(table: String): DeleteBuilder =
        DeleteBuilder(this, table)

    // ── Raw QAIL text protocol ──────────────────────────────────────

    /** Execute raw Qail DSL text. */
    suspend inline fun <reified T> query(dsl: String): QueryResponse<T> =
        request(HttpMethod.Post, "/qail") {
            setBody(dsl)
            contentType(ContentType.Text.Plain)
        }

    /** Execute raw Qail DSL via fast protocol (array-of-arrays). */
    suspend fun queryFast(dsl: String): FastQueryResponse =
        request(HttpMethod.Post, "/qail/fast") {
            setBody(dsl)
            contentType(ContentType.Text.Plain)
        }

    /** Execute a batch of Qail DSL queries. */
    suspend inline fun <reified T> batch(queries: List<String>): BatchResponse<T> =
        request(HttpMethod.Post, "/qail/batch") {
            setBody(mapOf("queries" to queries))
            contentType(ContentType.Application.Json)
        }

    // ── Utilities ───────────────────────────────────────────────────

    /** Health check. */
    suspend fun health(): HealthResponse = request(HttpMethod.Get, "/health")

    /** Get OpenAPI spec. */
    suspend fun openapi(): JsonObject = request(HttpMethod.Get, "/api/_openapi")

    /** Get schema introspection. */
    suspend fun schema(): JsonObject = request(HttpMethod.Get, "/api/_schema")

    /** Generate TypeScript interfaces from the gateway schema. */
    suspend fun generateTypes(): String = requestText(HttpMethod.Get, "/api/_schema/typescript")

    // ── Transactions ────────────────────────────────────────────────

    /** Start a new transaction session. */
    suspend fun beginTxn(): QailTxnSession {
        val res: TxnBeginResponse = request(HttpMethod.Post, "/txn/begin")
        return QailTxnSession(this, res.txnId)
    }

    // ── Realtime (WebSocket) ────────────────────────────────────────

    /**
     * Subscribe to a Postgres LISTEN channel via WebSocket.
     *
     * ```kotlin
     * val sub = qail.subscribe("orders") { payload ->
     *     println("Got: $payload")
     * }
     * // Later...
     * sub.unsubscribe()
     * ```
     */
    fun subscribe(
        channel: String,
        scope: CoroutineScope = CoroutineScope(Dispatchers.IO),
        onMessage: (String) -> Unit,
    ): QailSubscription {
        val sub = WebSocketSubscriptionImpl()

        sub.job = scope.launch {
            val token = config.token ?: config.tokenProvider?.invoke()
            val wsUrl = buildWebSocketUrl(token)

            try {
                client.webSocket({
                    url(wsUrl)
                    config.headers.forEach { (k, v) -> header(k, v) }
                    if (config.wsAuthMode == WebSocketAuthMode.HEADER && token != null) {
                        header(HttpHeaders.Authorization, "Bearer $token")
                    }
                }) {
                    // Send listen command
                    val cmd = kotlinx.serialization.json.buildJsonObject {
                        put("action", kotlinx.serialization.json.JsonPrimitive("listen"))
                        put("channel", kotlinx.serialization.json.JsonPrimitive(channel))
                    }
                    send(Frame.Text(json.encodeToString(kotlinx.serialization.json.JsonObject.serializer(), cmd)))

                    for (frame in incoming) {
                        if (!sub.active) break
                        if (frame is Frame.Text) {
                            try {
                                val msg = json.decodeFromString<JsonObject>(frame.readText())
                                val ch = msg["channel"]?.jsonPrimitive?.contentOrNull
                                val payloadElement = msg["payload"]
                                val payload = payloadElement?.jsonPrimitive?.contentOrNull ?: payloadElement?.toString()
                                if (ch == channel && payload != null) {
                                    onMessage(payload)
                                }
                            } catch (_: Exception) { }
                        }
                    }
                }
            } finally {
                sub.markClosed()
            }
        }

        return sub
    }

    // ── Internal HTTP ───────────────────────────────────────────────

    /** Internal typed request. */
    @PublishedApi
    internal suspend inline fun <reified T> request(
        method: HttpMethod,
        path: String,
        noinline block: HttpRequestBuilder.() -> Unit = {},
    ): T {
        val response = client.request("$baseUrl$path") {
            this.method = method

            // Default headers
            config.headers.forEach { (k, v) -> header(k, v) }

            // Auth token
            val token = config.token
                ?: config.tokenProvider?.invoke()
            token?.let { header(HttpHeaders.Authorization, "Bearer $it") }

            block()
        }

        if (!response.status.isSuccess()) {
            val text = response.bodyAsText()
            val body = try {
                json.decodeFromString<QailErrorBody>(text)
            } catch (_: Exception) {
                QailErrorBody(
                    code = "HTTP_${response.status.value}",
                    message = text,
                )
            }
            throw QailError(status = response.status.value, body = body)
        }

        return response.body()
    }

    /** Internal text request. */
    @PublishedApi
    internal suspend fun requestText(
        method: HttpMethod,
        path: String,
    ): String {
        val response = client.request("$baseUrl$path") {
            this.method = method
            config.headers.forEach { (k, v) -> header(k, v) }
            val token = config.token ?: config.tokenProvider?.invoke()
            token?.let { header(HttpHeaders.Authorization, "Bearer $it") }
        }
        if (!response.status.isSuccess()) {
            val text = response.bodyAsText()
            val body = try {
                json.decodeFromString<QailErrorBody>(text)
            } catch (_: Exception) {
                QailErrorBody(code = "HTTP_${response.status.value}", message = text)
            }
            throw QailError(status = response.status.value, body = body)
        }
        return response.bodyAsText()
    }

    @PublishedApi
    internal fun buildWebSocketUrl(token: String?): String {
        val wsBase = baseUrl
            .replace(Regex("^https"), "wss")
            .replace(Regex("^http"), "ws")
        val basePath = "$wsBase/ws"
        if (token == null || config.wsAuthMode != WebSocketAuthMode.QUERY) {
            return basePath
        }
        val sep = if (basePath.contains("?")) "&" else "?"
        val key = encodeQueryComponent(config.wsTokenQueryParam)
        val value = encodeQueryComponent(token)
        return "$basePath$sep$key=$value"
    }

    private fun encodeQueryComponent(value: String): String =
        URLEncoder.encode(value, StandardCharsets.UTF_8).replace("+", "%20")
}

/**
 * Handle for an active transaction session.
 */
class QailTxnSession(@PublishedApi internal val client: QailClient, val txnId: String) {

    /** Execute a query within this transaction. */
    suspend inline fun <reified T> query(dsl: String): QueryResponse<T> =
        client.request(HttpMethod.Post, "/txn/query") {
            header("X-Transaction-Id", txnId)
            setBody(dsl)
            contentType(ContentType.Text.Plain)
        }

    /** Commit the transaction. */
    suspend fun commit(): TxnEndResponse =
        client.request(HttpMethod.Post, "/txn/commit") {
            header("X-Transaction-Id", txnId)
        }

    /** Rollback the transaction. */
    suspend fun rollback(): TxnEndResponse =
        client.request(HttpMethod.Post, "/txn/rollback") {
            header("X-Transaction-Id", txnId)
        }

    /** Create or release a savepoint within this transaction. */
    suspend fun savepoint(action: String, name: String): SavepointResponse =
        client.request(HttpMethod.Post, "/txn/savepoint") {
            header("X-Transaction-Id", txnId)
            setBody(SavepointRequest(action = action, name = name))
            contentType(ContentType.Application.Json)
        }
}

File: sdk/swift/Sources/Qail/QailClient.swift
import Foundation

// MARK: - Configuration

/// Configuration for the Qail client.
public struct QailConfig: Sendable {
    /// Gateway base URL (e.g. `https://engine.example.com`).
    public let url: String
    /// JWT token or async token provider for refreshable auth.
    public let token: TokenSource?
    /// Additional default headers sent with every request.
    public let headers: [String: String]
    /// Request timeout in seconds (default: 30).
    public let timeout: TimeInterval
    /// Custom URLSession (defaults to `.shared`).
    public let session: URLSession
    /// Use snake_case ↔ camelCase key conversion (default: true).
    public let snakeCaseKeys: Bool

    public init(
        url: String,
        token: TokenSource? = nil,
        headers: [String: String] = [:],
        timeout: TimeInterval = 30,
        session: URLSession = .shared,
        snakeCaseKeys: Bool = true
    ) {
        self.url = url
        self.token = token
        self.headers = headers
        self.timeout = timeout
        self.session = session
        self.snakeCaseKeys = snakeCaseKeys
    }
}

/// Token source — either a static string or an async closure.
public enum TokenSource: Sendable {
    case string(String)
    case provider(@Sendable () async throws -> String)

    func resolve() async throws -> String {
        switch self {
        case .string(let t): return t
        case .provider(let fn): return try await fn()
        }
    }
}

// MARK: - Client

/// Qail Gateway client for iOS / macOS.
///
/// Provides fluent query builders, raw DSL execution, and utility
/// endpoints — mirroring the TypeScript SDK surface area.
///
/// ```swift
/// let qail = QailClient(config: .init(
///     url: "https://engine.example.com",
///     token: .string("my-jwt")
/// ))
///
/// // Fluent query
/// let users: [User] = try await qail.from("users")
///     .where("active", .eq, "true")
///     .limit(10)
///     .all()
///
/// // Raw DSL
/// let res = try await qail.query("get users fields id, name limit 10")
/// ```
public final class QailClient: Sendable {
    private let baseUrl: String
    private let defaultHeaders: [String: String]
    private let timeout: TimeInterval
    private let tokenSource: TokenSource?
    private let session: URLSession
    internal let decoder: JSONDecoder
    internal let encoder: JSONEncoder

    public init(config: QailConfig) {
        // Strip trailing slashes
        var url = config.url
        while url.hasSuffix("/") { url.removeLast() }
        self.baseUrl = url
        self.defaultHeaders = config.headers
        self.timeout = config.timeout
        self.tokenSource = config.token
        self.session = config.session

        let dec = JSONDecoder()
        let enc = JSONEncoder()
        if config.snakeCaseKeys {
            dec.keyDecodingStrategy = .convertFromSnakeCase
            enc.keyEncodingStrategy = .convertToSnakeCase
        }
        dec.dateDecodingStrategy = .iso8601
        enc.dateEncodingStrategy = .iso8601
        self.decoder = dec
        self.encoder = enc
    }

    // MARK: - Query builder entry points

    /// Start a SELECT query on a table.
    public func from<T: Decodable>(_ table: String) -> SelectBuilder<T> {
        SelectBuilder<T>(client: self, table: table)
    }

    /// Start an INSERT on a table.
    public func into<T: Decodable>(_ table: String) -> InsertBuilder<T> {
        InsertBuilder<T>(client: self, table: table)
    }

    /// Start an UPDATE on a table.
    public func update<T: Decodable>(_ table: String) -> UpdateBuilder<T> {
        UpdateBuilder<T>(client: self, table: table)
    }

    /// Start a DELETE on a table.
    public func delete(_ table: String) -> DeleteBuilder {
        DeleteBuilder(client: self, table: table)
    }

    // MARK: - Raw QAIL text protocol

    /// Execute raw Qail DSL text.
    ///
    /// ```swift
    /// let res = try await qail.query("get users fields id, name limit 10")
    /// ```
    public func query<T: Decodable>(_ dsl: String) async throws -> QueryResponse<T> {
        try await request(method: "POST", path: "/qail", body: dsl.data(using: .utf8), contentType: "text/plain")
    }

    /// Execute a batch of Qail DSL queries.
    public func batch<T: Decodable>(_ queries: [String]) async throws -> [BatchResult<T>] {
        let body = try JSONSerialization.data(withJSONObject: queries)
        return try await request(method: "POST", path: "/qail/batch", body: body)
    }

    // MARK: - Utilities

    /// Health check.
    public func health() async throws -> HealthResponse {
        try await request(method: "GET", path: "/health")
    }

    /// Get OpenAPI spec.
    public func openapi() async throws -> [String: Any] {
        try await requestRaw(method: "GET", path: "/api/_openapi")
    }

    /// Get schema introspection.
    public func schema() async throws -> [String: Any] {
        try await requestRaw(method: "GET", path: "/api/_schema")
    }

    /// Generate TypeScript interfaces from the gateway schema.
    ///
    /// Returns a string containing valid TypeScript interface declarations
    /// that can be written to a `.d.ts` file for type-safe queries.
    public func generateTypes() async throws -> String {
        try await requestText(method: "GET", path: "/api/_schema/typescript")
    }

    // MARK: - Realtime (WebSocket)

    /// Subscribe to a Postgres LISTEN channel via WebSocket.
    ///
    /// ```swift
    /// let sub = qail.subscribe("orders") { payload in
    ///     print("Got: \(payload)")
    /// }
    /// // Later...
    /// sub.unsubscribe()
    /// ```
    public func subscribe(channel: String, onMessage: @escaping @Sendable (String) -> Void) -> QailSubscription {
        let wsUrl = baseUrl.replacingOccurrences(of: "http", with: "ws") + "/ws"
        let url = URL(string: wsUrl)!
        let task = session.webSocketTask(with: url)
        let subscription = WebSocketSubscription(task: task, channel: channel, onMessage: onMessage)
        task.resume()

        // Send listen command once connected
        let listenMsg = #"{"action":"listen","channel":"\#(channel)"}"#
        task.send(.string(listenMsg)) { _ in }

        // Start receiving
        subscription.startReceiving()
        return subscription
    }

    // MARK: - Raw HTTP (for Workers endpoints)

    /// Make a typed request to any path (e.g. Workers endpoints).
    ///
    /// ```swift
    /// struct LoginResponse: Decodable { let token: String }
    /// let res: LoginResponse = try await qail.post("/public/auth/login", body: creds)
    /// ```
    public func get<T: Decodable>(_ path: String) async throws -> T {
        try await request(method: "GET", path: path)
    }

    public func post<T: Decodable, B: Encodable>(_ path: String, body: B) async throws -> T {
        let data = try encoder.encode(body)
        return try await request(method: "POST", path: path, body: data)
    }

    public func patch<T: Decodable, B: Encodable>(_ path: String, body: B) async throws -> T {
        let data = try encoder.encode(body)
        return try await request(method: "PATCH", path: path, body: data)
    }

    public func post<T: Decodable>(_ path: String) async throws -> T {
        try await request(method: "POST", path: path)
    }

    public func put<T: Decodable, B: Encodable>(_ path: String, body: B) async throws -> T {
        let data = try encoder.encode(body)
        return try await request(method: "PUT", path: path, body: data)
    }

    public func put<T: Decodable>(_ path: String) async throws -> T {
        try await request(method: "PUT", path: path)
    }

    public func delete<T: Decodable>(_ path: String) async throws -> T {
        try await request(method: "DELETE", path: path)
    }

    public func delete<T: Decodable, B: Encodable>(_ path: String, body: B) async throws -> T {
        let data = try encoder.encode(body)
        return try await request(method: "DELETE", path: path, body: data)
    }

    /// PUT with no response body expected.
    public func put(_ path: String) async throws {
        _ = try await performRequest(method: "PUT", path: path)
    }

    /// PUT with body, no response body expected.
    public func put<B: Encodable>(_ path: String, body: B) async throws {
        let data = try encoder.encode(body)
        _ = try await performRequest(method: "PUT", path: path, body: data)
    }

    /// DELETE with no response body expected.
    public func delete(_ path: String) async throws {
        _ = try await performRequest(method: "DELETE", path: path)
    }

    // MARK: - Internal HTTP

    /// Internal JSON request that returns a Decodable type.
    internal func request<T: Decodable>(
        method: String,
        path: String,
        body: Data? = nil,
        contentType: String = "application/json"
    ) async throws -> T {
        let data = try await performRequest(method: method, path: path, body: body, contentType: contentType)
        return try decoder.decode(T.self, from: data)
    }

    /// Internal request that returns raw JSON dictionary.
    private func requestRaw(
        method: String,
        path: String
    ) async throws -> [String: Any] {
        let data = try await performRequest(method: method, path: path)
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw QailError(code: "PARSE_ERROR", message: "Response is not a JSON object", details: nil, requestId: nil, hint: nil, table: nil, column: nil)
        }
        return json
    }

    /// Internal request that returns plain text.
    internal func requestText(
        method: String,
        path: String
    ) async throws -> String {
        let data = try await performRequest(method: method, path: path)
        guard let text = String(data: data, encoding: .utf8) else {
            throw QailError(code: "PARSE_ERROR", message: "Response is not valid UTF-8", details: nil, requestId: nil, hint: nil, table: nil, column: nil)
        }
        return text
    }

    /// Core HTTP engine.
    private func performRequest(
        method: String,
        path: String,
        body: Data? = nil,
        contentType: String = "application/json"
    ) async throws -> Data {
        guard let url = URL(string: "\(baseUrl)\(path)") else {
            throw QailError(code: "INVALID_URL", message: "Bad URL: \(baseUrl)\(path)", details: nil, requestId: nil, hint: nil, table: nil, column: nil)
        }

        var request = URLRequest(url: url, timeoutInterval: timeout)
        request.httpMethod = method
        request.setValue(contentType, forHTTPHeaderField: "Content-Type")

        // Default headers
        for (key, value) in defaultHeaders {
            request.setValue(value, forHTTPHeaderField: key)
        }

        // Auth token
        if let tokenSource {
            let token = try await tokenSource.resolve()
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        // Body
        if method != "GET", let body {
            request.httpBody = body
        }

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw QailError(code: "NETWORK_ERROR", message: "Invalid response", details: nil, requestId: nil, hint: nil, table: nil, column: nil)
        }

        guard (200..<300).contains(httpResponse.statusCode) else {
            throw QailError.from(data: data, status: httpResponse.statusCode)
        }

        return data
    }
}
